### PR TITLE
PODA-7: Update the LTI 1.3 Library to cache auth tokens

### PR DESCRIPTION
**Pull request by cophaso**
## Summary of Changes

Implements Guzzle and caches auth tokens. 

## Testing

- [x] I have run `composer test`
- [x] I have run `composer lint-fix` 
- In `backend/app-api/composer.json` I set  `"packbackbooks/lti-1p3-tool": "dev-cache-auth-tokens",` to point the BE to a specific library branch. 
  - then run `composer install` and `make pull` 
- Then seeded GBS Data `php artisan db:seed --class=GradebookSyncDataSeeder`
- launch an assignment. 
  - Log into canvas using a user from the GBS Dataseeder and open up an assignment, make sure the user is associated to a packback user.
- Exec'd into docker container (`docker-compose exec app-api-php-fpm bash`) and opned a tinker session (`php artisan tinker`)
- Then queued grades and synced 

```
foreach(Grade::where('community_id', 'c0000012-0000-8000-0000-000000000000')->get() as $grade) if ($grade->canBeSynced) SyncGradesForLtiKey::queueGradeToSync($grade);

foreach (KeyDetails::get() as $key) Queue::push(new SyncGradesForLtiKey($key->id));
```
Was able to sync grades without an error about the access token being expired or a bad request from the guzzle client. 

I did run into this error a lot when syncing grades

```
{\"errors\":{\"type\":\"unprocessable_entity\",\"message\":\"User not found in course or is not a student\"}}
today at 12:49:57 PM ","lti_key":7}
```

[It's a normal error ](https://packbackco.slack.com/archives/C01L8QXBJF4/p1623968265229000) seems to happen if a user is linked to packback correctly. 

## PR Dependencies

- [ ] https://github.com/packbackbooks/questions/pull/4455

### CODE REVIEW

File: src/LtiAssignmentsGradesService.php [hunk 1]
```diff
@@ -41,7 +41,7 @@ public function putGrade(LtiGrade $grade, LtiLineitem $lineitem = null)
             $this->service_data['scope'],
             LtiServiceConnector::METHOD_POST,
             $score_url,
-            strval($grade),
+            $grade,
```
Comment (dbhynds) [inline src/LtiAssignmentsGradesService.php:44]: What is the purpose of these changes?

File: src/LtiServiceConnector.php [hunk 2]
```diff
@@ -13,90 +15,99 @@ class LtiServiceConnector implements LtiServiceConnectorInterface
...
-        $resp = curl_exec($ch);
-        $token_data = json_decode($resp, true);
-        curl_close($ch);
-
-        return $this->access_tokens[$scope_key] = $token_data['access_token'];
+        $url = $this->registration->getAuthTokenUrl();
+
+        // Get Access
+        $response = $this->client->post($url, [
+            'timeout' => 10,
```
Comment (dbhynds) [inline src/LtiServiceConnector.php:71]: Rather than setting this in the library, is there a way that we can configure it on the client that gets passed into it (and therefore, within our app instead of this library)?

File: tests/LtiServiceConnectorTest.php [hunk 3]
```diff
@@ -12,8 +14,10 @@ class LtiServiceConnectorTest extends TestCase
     public function testItInstantiates()
     {
         $registration = Mockery::mock(LtiRegistrationInterface::class);
+        $cache = Mockery::mock(Cache::class);
```
Comment (dbhynds) [inline tests/LtiServiceConnectorTest.php:17]: Now that we're not relying on curl, can we add some additional automated tests here?

File: src/LtiServiceConnector.php [hunk 4]
```diff
@@ -13,90 +15,99 @@ class LtiServiceConnector implements LtiServiceConnectorInterface
...
+            'Authorization' => 'Bearer '.$this->getAccessToken($scopes),
+            'Accept' => $accept,
         ];
-        curl_setopt($ch, CURLOPT_URL, $url);
-        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
-        curl_setopt($ch, CURLOPT_HEADER, 1);
-        curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
-        curl_setopt($ch, CURLOPT_TIMEOUT, 60);
+
         if ($method === 'POST') {
```
Comment (dbhynds) [inline src/LtiServiceConnector.php:90]: This seems like the ideal use case for a `case switch`

File: src/LtiServiceConnector.php [hunk 5]
```diff
@@ -13,90 +15,99 @@ class LtiServiceConnector implements LtiServiceConnectorInterface
...
 
         // Build up JWT to exchange for an auth token
-        $client_id = $this->registration->getClientId();
-        $jwt_claim = [
-                'iss' => $client_id,
-                'sub' => $client_id,
+        $clientId = $this->registration->getClientId();
+
+        // Store access token with a unique key
+        $accessTokenKey = $scopeKey.'-'.$clientId;
```
[the remark this answers is not in the document]
Comment (dbhynds) [reply src/LtiServiceConnector.php:40]: And actually, semi-related... for the sake of testing, it might be easiest to add a `getAccessTokenCacheKey()` function to this class.

**Review by dbhynds — CHANGES_REQUESTED** [review]

File: src/LtiAssignmentsGradesService.php (hunk 1 above)
Comment (cophaso) [reply src/LtiAssignmentsGradesService.php:44]: Guzzle expects a json object in the body. I was getting these errors because we were passing along a json blob converted into a string value.

<details>

Screenshot from 2021-06-22 10-54-41


</details>

File: src/LtiServiceConnector.php (hunk 2 above)
Comment (cophaso) [reply src/LtiServiceConnector.php:71]: Would that work when timeout varies? For the auth token it's 10 (seconds?) but to make the service request timeout is 60. I'm assuming you mean to set a default timeout on the client?

File: tests/LtiServiceConnectorTest.php (hunk 3 above)
Comment (cophaso) [reply tests/LtiServiceConnectorTest.php:17]: Are these automated tests? they feel more like unit tests? Regardless, mocking the function calls may get a bit hairy. Might need to possibly pair.

File: src/LtiServiceConnector.php (hunk 2 above)
Comment (dbhynds) [reply src/LtiServiceConnector.php:70]: Hmm. Why is the auth token timeout 10 seconds, when the service request is 60? Where did you get those numbers from?

From my experience (and based on Canvas's rate limiting equation), the auth request is 5x slower than the service request on average (1.5s vs 0.3s).

I'd think we'd want to let whoever is using this request configure the timeout for the whole client, rather than forcing them to use the timeouts we decide. That does have the downside tho of not being able to vary the timeout between the auth and service request. My personal opinion is that it's fine to accept that downside, but I might not be understanding the full context.

File: src/LtiServiceConnector.php [hunk 6]
```diff
@@ -110,4 +111,11 @@ public function makeServiceRequest(array $scopes, $method, $url, $body = null, $
             'body' => json_decode($respBody, true),
         ];
     }
+
+    private function getAccessTokenCacheKey(array $scopes)
+    {
+        $scopeKey = md5(implode('|', $scopes));
```
Comment (dbhynds) [inline src/LtiServiceConnector.php:117]: Let's move the `sort($scopes)` in here too.

File: src/LtiServiceConnector.php [hunk 7]
```diff
@@ -87,19 +86,21 @@ public function makeServiceRequest(array $scopes, $method, $url, $body = null, $
...
-                'headers' => $headers,
-                'json' => $body,
-                'timeout' => 60,
-            ]);
-        } else {
-            $response = $this->client->request($method, $url, [
-                'timeout' => 60,
-                'headers' => $headers,
-            ]);
+        switch ($method) {
```
Comment (dbhynds) [inline src/LtiServiceConnector.php:89]: It might be worth converting the string to uppercase, so that it's not case sensitive.

File: src/LtiServiceConnector.php (hunk 2 above)
Comment (cophaso) [reply src/LtiServiceConnector.php:70]: ah looks like the 10 & 60 second timeouts set in this file are arbitrary. Finally able to figure this out. so I removed the timeouts and created a client in GradebookSync13 and SendGradeBack with a timeout. I set it to 60 seconds.

https://github.com/packbackbooks/questions/pull/4455/files#diff-9e91242e398250a7a8bda45567746be7739b930668d09a87b10fc88562fee1d8R31-R34

File: tests/LtiServiceConnectorTest.php (hunk 3 above)
Comment (cophaso) [reply tests/LtiServiceConnectorTest.php]: Unit testing this was a head scratcher. I added a test to make sure that a cached token could be retrieved. Just to get this moving, I think that's okay for now. We didn't previously have a test for these methods. Something to look into is how we can add resources to better test this library.

**Review by lin-brian-l — APPROVED** [review]
@cophaso I didn't do a thorough review of the code, but I've verified that Davo's feedback was addressed and was able to QA this as per your instructions - great work!
